File: FE/DATN-Fashion-Shop-FE/src/app/models/Product/ProductVariant.ts
import { BaseEntity } from '../BaseEntity';
import { AttributeValue} from './AttributeValue';
import { Product} from './product';
import { ProductMedia} from './ProductMedia';

export class ProductVariant extends BaseEntity {
  id: number;
  salePrice: number;
  colorValue?: AttributeValue;
  sizeValue?: AttributeValue;
  product: Product;
  productMedias: ProductMedia[];

  constructor(data?: Partial<ProductVariant>) {
    super(data);
    this.id = data?.id || 0;
    this.salePrice = data?.salePrice || 0;
    this.colorValue = data?.colorValue ? new AttributeValue(data.colorValue) : undefined;
    this.sizeValue = data?.sizeValue ? new AttributeValue(data.sizeValue) : undefined;
    this.product = data?.product ? new Product(data.product) : new Product();
    this.productMedias = data?.productMedias?.map(pm => new ProductMedia(pm)) || [];
  }

  getAdjustedPrice(): number {
    if (
      this.product.promotion &&
      this.product.promotion.isActive &&
      new Date() >= new Date(this.product.promotion.startDate) &&
      new Date() <= new Date(this.product.promotion.endDate)
    ) {
      const discountPercentage = this.product.promotion.discountPercentage;
      return this.salePrice - (this.salePrice * discountPercentage) / 100;
    }
    return this.salePrice;
  }
}

File: DATN-Fashion-Shop-BE/src/main/java/com/example/DATN_Fashion_Shop_BE/model/ProductVariant.java
package com.example.DATN_Fashion_Shop_BE.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "product_variants")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ProductVariant extends BaseEntity{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "sale_price", nullable = false)
    private Double salePrice;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "color_value_id")
    private AttributeValue colorValue;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "size_value_id")
    private AttributeValue sizeValue;

    @ManyToOne
    @JoinColumn(name = "product_id")
    private Product product;

    @ManyToMany(mappedBy = "productVariants")
    private Set<ProductMedia> productMedias = new HashSet<>();

    public Double getAdjustedPrice() {
        if (product.getPromotion() != null &&
                product.getPromotion().getIsActive() &&
                LocalDateTime.now().isAfter(product.getPromotion().getStartDate()) &&
                LocalDateTime.now().isBefore(product.getPromotion().getEndDate())) {
            Double discountPercentage = product.getPromotion().getDiscountPercentage();
            return salePrice - (salePrice * discountPercentage / 100);
        }
        return salePrice;
    }
}
